builder/builder.go

```diff
@@ -132,3 +207,76 @@ func Receipt(fee *felt.Felt, feeUnit core.FeeUnit, txHash *felt.Felt, trace *vm.
 		RevertReason:       trace.RevertReason(),
 	}
 }
+
+func (b *Builder) listenPool(ctx context.Context) error {
+	for {
+		if err := b.depletePool(ctx); err != nil {
+			if !errors.Is(err, db.ErrKeyNotFound) {
+				return err
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-b.pool.Wait():
+			continue
+		}
+	}
+}
+
+func (b *Builder) depletePool(ctx context.Context) error {
+	for {
+		userTxn, err := b.pool.Pop()
+		if err != nil {
+			return err
+		}
+		b.log.Debugw("running txn", "hash", userTxn.Transaction.Hash().String())
+
+		if err = b.runTxn(&userTxn); err != nil {
+			var txnExecutionError vm.TransactionExecutionError
+			if !errors.As(err, &txnExecutionError) {
+				return err
+			}
+
+			b.log.Debugw("failed txn", "hash", userTxn.Transaction.Hash().String(), "err", err.Error())
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+	}
+}
+
+func (b *Builder) runTxn(txn *mempool.BroadcastedTransaction) error {
+	b.pendingLock.Lock()
+	defer b.pendingLock.Unlock()
+
+	state := blockchain.NewPendingState(b.pendingBlock.StateUpdate.StateDiff, b.pendingBlock.NewClasses, b.headState)
+	var classes []core.Class
+	if txn.DeclaredClass != nil {
+		classes = append(classes, txn.DeclaredClass)
+	}
+
+	fee, trace, err := b.vm.Execute([]core.Transaction{txn.Transaction}, classes, b.pendingBlock.Block.Number,
```

Comment:
vm.Execute() updates the state right? Think we should add a comment here?

Reply:
Yeap, it does. What is it that you want clarified in the comment exactly?

Reply:
Maybe builder isn't the best place to have this comment, but imo, it would be useful to have in the definition of Execute() (vm/vm.go)